src/service_discovery/mod.rs

```diff
@@ -149,17 +151,11 @@ impl ServiceDiscovery {
         let our_current_listeners = unwrap!(self.our_listeners.lock()).iter().cloned().collect();
         let resp = DiscoveryMsg::Response(our_current_listeners);
 
-        let serialised_resp = serialise(&resp)?;
-
         if let Some(peer_addr) = self.reply_to.pop_front() {
-            match self.socket.send_to(&serialised_resp[..], &peer_addr) {
+            match self.socket.write_to(Some((resp, peer_addr, 0))) {
                 // UDP is all or none so if anything is written we consider it written
-                Ok(_bytes_send) => (),
-                Err(ref e)
-                    if e.kind() == ErrorKind::Interrupted || e.kind() == ErrorKind::WouldBlock =>
-                {
-                    self.reply_to.push_front(peer_addr)
-                }
+                Ok(true) => (),
+                Ok(false) => self.reply_to.push_front(peer_addr),
```

Comment:
Do you need to do this book-keeping ? `UdpSock` should have internally buffered it for a resent when the readiness for socket fires again.